### ISSUE TITLE
Token generation mismatch: hidden input does not match the session variable

### ISSUE THREAD

**Issue by jamespsterling** [paragonie/anti-csrf#12]
Inputs on form,
```
<!--
--><input type="hidden" name="_CSRF_INDEX" value="v6Dzi3KrRDV68kNdPFCES+UU"><!--
--><input type="hidden" name="_CSRF_TOKEN" value="wLoNJygvlKTxBEuhTa/WCjnvtoYldgmTet7MsFQlXU0=">
```

Session variables dumped at the end of page, 
```
array(1) {
  ["CSRF"]=&gt;
  array(1) {
    ["v6Dzi3KrRDV68kNdPFCES+UU"]=&gt;
    array(4) {
      ["created"]=&gt;
      int(20160114152843)
      ["uri"]=&gt;
      string(1) "/"
      ["token"]=&gt;
      string(44) "T0kXM8I9nzUFv3w7flJTlbOjFa1OEMNR+5xwnHvpqr4="
      ["lockto"]=&gt;
      string(5) "login"
    }
  }
}
```

**Comment by paragonie-scott** [comment]
What is the value of `protected $hmac_ip`?

**Comment by ncou** [comment]
Hi,

Same thing here. the token set in session is different from the value in the hidden text field.
The _CSRF_INDEX is the same in session and in the hidden text field.

The validateRequest return "TRUE" even if the value in the field is different from the value session.

$hmac_ip is TRUE

**Comment by ncou** [comment]
After a quick debug, it's because hmac_ip is TRUE.

If set to false, the token value in the hidden field is the same than in the session.

**Comment by paragonie-scott** [comment]
Yep. What this feature does is, instead of just placing the CSRF token in the form output, it outputs `hash_hmac('sha256', $IPaddress, $csrfToken)`.

https://github.com/paragonie/anti-csrf/blob/606274f8f8c6aa0a807e656d7fc0603b5c78fdac/src/AntiCSRF.php#L162-L174

I'm tagging this as a documentation bug.

**Comment by amolbarewar** [comment]
I am also facing the same issue.
 <input type="hidden" name="_CSRF_INDEX" value="cmYUHdfdMdp02CflvuizLOOQ" /><input type="hidden" name="_CSRF_TOKEN" value="pN+bcGTy4ZPo/T5/ITHonKVVEkOshpn966DbRcRFSUU=" />


CSRF token did not match with the value in the session object. $hmac_ip is FALSE in this case.

$_SEESION = 
array(1) {
  ["CSRF"]=>
  array(11) {
    ["fWYvREJG//jYOtz64O0VLW73"]=>
    array(4) {
      ["created"]=>
      int(20160729114830)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "O+p6Kl2jRK/5DX3pvkyBDU+1M/WRah7izCKF4zpm1tQ="
      ["lockto"]=>
      string(22) "createaccount-exec.php"
    }
    ["Axloe4OegP5C8P8XSH87VvY1"]=>
    array(4) {
      ["created"]=>
      int(20160729114913)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "YC8NGEBZo5hEyLDPT5jve35mx3yXBx5eP8RditxgNDs="
      ["lockto"]=>
      string(22) "createaccount-exec.php"
    }
    ["xRLieoLmedDRtIQjNSa2s3iJ"]=>
    array(4) {
      ["created"]=>
      int(20160729115540)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "gbvpbevc3b6uiZYzLoGyzzIlgCHkOydmwTWfKGoK9fd7"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["9WWmQ3p8/4S+gNTqdk/m1u3Z"]=>
    array(4) {
      ["created"]=>
      int(20160729115544)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "xuxZxqG6inzdmf6BT8TvviI7k5D935zw5GrrKOIDGtC1"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["y1G+xXzPR6gJKsRTOqfU6T8E"]=>
    array(4) {
      ["created"]=>
      int(20160729121304)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "S4tmoRSvLk+8EKBEV/LiPpC2b8UlETlEvALmKPbXYqbp"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["9Usso888KcBSHnuVZLSPzW0V"]=>
    array(4) {
      ["created"]=>
      int(20160729121307)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "WkV6Sq2FyQXgSPwXs+syiXLXYFWP6k7PZhmng/obvZhh"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["aGI7RF/tTdRjDBl1JZmMZvUC"]=>
    array(4) {
      ["created"]=>
      int(20160729121939)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "SkEX9Y4LlmSDCIv/08Ad+CpaDHaSmEcC3CseORf5E286"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["5mc8GsJwuCczRwK9VtnR/T87"]=>
    array(4) {
      ["created"]=>
      int(20160729122141)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "29NSxkYMWhR9iXzLPfQKpsjKwPKR+4L04sPr5x9eOXOz"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["VZKaKKTKesEOsXMEXAGTxEXR"]=>
    array(4) {
      ["created"]=>
      int(20160801094524)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "Btzl1pAvWDUWsuv9upMK/OlugS1HB2Id90d6+4ddgZxe"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["rflTVbF4UAyYrWjqptaAyBoU"]=>
    array(4) {
      ["created"]=>
      int(20160801094526)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "zjZ5PJ/jJXX4Z7l2crR7yBLVnutkDP8PPZXPmlDenZV6"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
    ["0YfVEUvZvtvw79sZfLMoV6Dk"]=>
    array(4) {
      ["created"]=>
      int(20160801094527)
      ["uri"]=>
      string(1) "/"
      ["token"]=>
      string(44) "3GdcVUw9mMEZ//STt6HSsSXB6+aJyUAesmnpoby7tLIr"
      ["lockTo"]=>
      string(22) "createaccount-exec.php"
    }
  }
}

**Comment by amolbarewar** [comment]
Aftre a debug I came to know that $this->session[$this->sessionIndex] is not set in constuctor. while in other backend file I can see valid value for $_SESSION
